examples/typescript-interop/node/demo.c

```diff
@@ -0,0 +1,402 @@
+    // https://nodejs.org/api/n-api.html#napi_get_value_string_utf8
+    size_t capacity = len + 1;
+
+    // Create a RocStr and write it into the out param
+    if (capacity < sizeof(struct RocStr))
+    {
+        // If it can fit in a small string, use the string itself as the buffer.
+        // First, zero out those bytes; small strings need to have zeroes for any bytes
+        // that are not part of the string, or else comparisons between small strings might fail.
+        *roc_str = empty_roc_str();
+
+        // This writes the actual number of bytes copied into len. Theoretically they should be the same,
+        // but it could be different if the buffer was somehow smaller. This way we guarantee that
+        // the RocStr does not present any memory garbage to the user.
+        status = napi_get_value_string_utf8(env, node_string, (char*)roc_str, sizeof(struct RocStr), &len);
+
+        if (status != napi_ok)
+        {
+            return status;
+        }
+
+        // We have to write the length into the buffer *after* Node copies its bytes in,
+        // because Node will have written a null terminator, which we may need to overwrite.
+        write_small_str_len(len, roc_str);
+    }
+    else
+    {
+        // capacity was too big for a small string, so make a heap allocation and write into that.
+        uint8_t *buf = (uint8_t*)roc_alloc(capacity, __alignof__(char));
+
+        // This writes the actual number of bytes copied into len. Theoretically they should be the same,
+        // but it could be different if the buffer was somehow smaller. This way we guarantee that
+        // the RocStr does not present any memory garbage to the user.
+        status = napi_get_value_string_utf8(env, node_string, (char*)buf, capacity, &len);
+
+        if (status != napi_ok)
+        {
+            // Something went wrong, so free the bytes we just allocated before returning.
+            roc_dealloc((void *)&buf, __alignof__(char *));
+
+            return status;
+        }
+
+        *roc_str = roc_str_init_large(buf, len, capacity);
+    }
+
+    return status;
+}
+
+// Consume the given RocStr (decrement its refcount) after creating a Node string from it.
+napi_value roc_str_into_node_string(napi_env env, struct RocStr roc_str) {
+    bool is_small = is_small_str(roc_str);
+    char* roc_str_contents;
+
+    if (is_small)
+    {
+        // In a small string, the string itself contains its contents.
+        roc_str_contents = (char*)&roc_str;
+    }
+    else
+    {
+        roc_str_contents = (char*)roc_str.bytes;
+    }
+
+    napi_status status;
+    napi_value answer;
+
+    status = napi_create_string_utf8(env, roc_str_contents, roc_str_len(roc_str), &answer);
+
+    if (status != napi_ok)
+    {
+        answer = NULL;
+    }
+
+    // Decrement the RocStr because we consumed it.
+    if (!is_small)
+    {
+        decref((void *)&roc_str.bytes, __alignof__(uint8_t *));
```

Comment:
I guess one more comment here. Not sure if you want to change the api of decref to deal with it or deal with it here. For a seamless slice of a Str to be freed. If the `roc_str.length` is negative as an `isize` (first bit set), instead of using `roc_str.bytes` here, you would use `roc_str.capacity << 1`

Reply:
Gotcha - how's this look? 6845dae2b2d48dd5d2451d780a2251f37c5704fc

Reply:
That should all work, though `decref_list` is technically a wrong name and is just the inner `decref`. If you wanted a `decref_list` that takes in a RocList, it would need to check if the capacity is negative. Then do the same thing as done with RocStr. The bytes pointer would would be `roc_list.capacity << 1` So maybe change the name back from `decref_list` to `decref`

Reply:
Good call! Done in dfcf08691089703efb218750a3f842d4e74745cf